### PR TITLE
Extend_filtration when all the vertices have the same filtration value

### LINKED ISSUE #887: Simplex_tree extend_filtration returns `NaN` when all the vertices have the same filtration value
## Example that fails
```python
from gudhi import SimplexTree as STree
import numpy as np

st.insert_batch(np.array([[6, 1, 5]]), np.array([4.0, 4.0, 4.0]))
list(st.get_simplices())
# [([1], 4.0), ([5], 4.0), ([6], 4.0)]
st.extend_filtration()
list(st.get_simplices())
# [([1, 7], nan), ([1], nan), ([5, 7], nan), ([5], nan), ([6, 7], nan), ([6], nan), ([7], -3.0)]
```

## Reason
https://github.com/GUDHI/gudhi-devel/blob/23431ea266e6ba6d8a1926707fadfe31bb7db904/src/Simplex_tree/include/gudhi/Simplex_tree.h#L1689 returns `NaN` when `minval = maxval`

## Proposal
An exception could be thrown when detected

### PR DESCRIPTION
**Pull request by VincentRouvreau**
Fix #887 
Small code modification for `extend_filtration` and `decode_extended_filtration`
Some unitary tests  for `extend_filtration` and `decode_extended_filtration`

### CODE REVIEW

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 1]
```diff
@@ -1781,43 +1779,40 @@ class Simplex_tree {
...
     Simplex_tree st_copy = *this;
 
     // Add point for coning the simplicial complex
     this->insert_simplex_raw({maxvert}, -3);
 
-    // For each simplex
-    std::vector<Vertex_handle> vr;
-    for (auto sh_copy : st_copy.complex_simplex_range()){
+    Filtration_value scale = maxval-minval;
+    if (std::fabs(maxval - minval) > std::numeric_limits<Filtration_value>::epsilon())
```
Comment (mglisse) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h:1798]: Comparing a value to epsilon does not make sense to me. epsilon is for relative stuff, or bit manipulations.

Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree.h]: Done on [8550ad3](https://github.com/GUDHI/gudhi-devel/pull/888/commits/8550ad3b01f047a30044cbcce8c142a421da96f4)

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 2]
```diff
@@ -1781,43 +1779,40 @@ class Simplex_tree {
...
+      scale = 1 / scale;
 
-      // Locate simplex
-      vr.clear();
-      for (auto vh : st_copy.simplex_vertex_range(sh_copy)){
-        vr.push_back(vh);
-      }
+    // For each simplex
+    for (auto sh_copy : st_copy.complex_simplex_range()) {
+      std::vector<Vertex_handle> vr(st_copy.simplex_vertex_range(sh_copy).begin(), st_copy.simplex_vertex_range(sh_copy).end());
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree.h]: I used the syntax you are proposing on [0ed2a8c](https://github.com/GUDHI/gudhi-devel/pull/888/commits/0ed2a8c2417385098c33035e41e9fefe86bb8c07)

File: src/Simplex_tree/test/simplex_tree_extended_filtration_unit_test.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  BOOST_CHECK(std::isnan(filt));
+  BOOST_CHECK(est == Gudhi::Extended_simplex_type::EXTRA);
+
+  std::tie(filt, est) = st.decode_extended_filtration(Filtration_value(2.1), efd);
+  std::clog << filt << " - " << static_cast<int>(est) << std::endl;
+  BOOST_CHECK(std::isnan(filt));
+  BOOST_CHECK(est == Gudhi::Extended_simplex_type::EXTRA);
+}
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(simplex_tree_extend_same_filtration, Stree, list_of_tested_variants) {
```
Comment (mglisse) [inline src/Simplex_tree/test/simplex_tree_extended_filtration_unit_test.cpp:169]: Could we add some calls to `decode` in this test?

Comment (VincentRouvreau) [reply src/Simplex_tree/test/simplex_tree_extended_filtration_unit_test.cpp:169]: I added some tests on [025f140](https://github.com/GUDHI/gudhi-devel/pull/888/commits/025f14017de37998782aa90b7d3ea9bf0015617a)

**Review by mglisse — APPROVED** [review]

**Pull request merged**